### PR TITLE
04 コメント機能の実装

### PR DESCRIPTION
**Pull request by miketa-webprgr**
## 作業ノート
- [Rails特訓コース Issue04ノート](https://github.com/miketa-webprgr/TIL/blob/master/11_Rails_Intensive_Training/04_issue_note.md)
-  [Rails特訓コース Issue04ノート(CRUD実装に限る）](https://github.com/miketa-webprgr/TIL/blob/master/11_Rails_Intensive_Training/04_issue_note_comments-crud-ajax.md)

## 実装内容

- 投稿に対するコメントのCRUD機能を実装
- 編集・更新はモーダルを表示させ非同期で行う
- shallowルーティングを使用する
- Commentモデルに対して文字列長などのバリデーションを付与
- Commentモデルの国際化
- 気まぐれでswearjarを実装（コメントにNGワードを含むことを禁じた）

## 確認方法
1. `git clone https://github.com/miketa-webprgr/instagram_clone.git`
2. `git checkout git checkout -b feature/01_login_logout origin/feature/01_login_logout`
3. `bundle install`
4. `yarn install`
5. `MySQL と Redis を立ち上げる`
6. `rails db:migrate`
7. `rails db:seed`

## コメント
- パーシャル化がされていると、繋がりを追うのが大変だった
- JSのコードを理解するのが大変だった（しかもslim内にrubyと入り混じり、最初はどっちなのかわからないこともあった）
- 「○っぱい」ってコメント投稿できないようにするために徹夜していたとき、ふと我に返って悲しくなった

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    # comments.buildなどのメソッドはRailsガイドを参照
+    # https://railsguides.jp/association_basics.html#has-manyで追加されるメソッド-collection-build-attributes
+
+    # current_user.comments.buildにて以下が生成される
+    # {id: nil, body: nil, user_id: 40, post_id: nil, created_at: nil, updated_at: nil}
+    # ここに「body」と「post_id」を持ってきてあげればよい
+    # よって、comment_paramsメソッドでは、:bodyだけpermitして、そこにpost_idを付与してあげるのがベストの形となる
+
+    # ----- 質問 -----
+    # だいそんさんの書き方だと、buildとsaveの２行で分けていましたが、createでも大丈夫でしょうか？
```
Comment (DaichiSaito) [inline app/controllers/comments_controller.rb:14]: OKです！

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    if ng_words?(comment_update_params[:body])
+      @ng_flag = true
+    else
+      @comment.update(comment_update_params)
+    end
+  end
+
+  def destroy
+    @comment = current_user.comments.find(params[:id])
+    # 仕組み上、通常の操作をしている限りエラーになるはずがないので、destroyに失敗した場合は例外処理とする
```
Comment (DaichiSaito) [inline app/controllers/comments_controller.rb:38]: 👍

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    params.require(:comment).permit(:body).merge(post_id: params[:post_id])
+  end
+
+  # commentをupdateする場合、既にpostへ紐づいたcommentのbodyを差し替えるだけなので、bodyのみをpermitすればよい
+  def comment_update_params
+    params.require(:comment).permit(:body)
+  end
+
+  # NGワードが含まれているかチェックするメソッド
+  # 本当はモデルでバリデーションをかけたくて行錯誤したが上手くいかず（時間をかけたくないので、適当なところで切り上げ）
```
Comment (DaichiSaito) [inline app/controllers/comments_controller.rb:56]: そうっすねー。モデルでカスタムバリデーション使うのが良いかもなー。
`@ng_flag`ってのもあまりかっこよくはないですよねー。

Comment (miketa-webprgr) [reply app/controllers/comments_controller.rb:56]: やっぱりカスタムバリデーションですか。。。

最初はそれがいいかなとは思ったんですけど、あまり上手くいかず、コントローラの方でやる方向に
切り替えちゃいました。@ng_flagもやっぱり微妙ですか。。。

ただ、何も状態から自分の脳味噌だけ使って書くと、コードをどう書くべきか色々と考えさせられますね。
眠くて、妥協に妥協を重ねていっちゃいましたけど笑

自分の脳味噌だけで書いて、みんなでコードを見せ合う会がやれると勉強になりそうだなと思いました！

File: app/models/post.rb [hunk 4]
```diff
@@ -27,4 +27,8 @@ class Post < ApplicationRecord
   validates :images, presence: true
 
   belongs_to :user
+  # ふと気になったが、destroyオプションとdeleteオプションの違いについても調べてみた
+  # deleteの場合、子であるcommentsまで削除されないので要注意
+  # [delete, delete\_all, destroy, destroy\_allについて \- Qiita](https://qiita.com/kamelo151515/items/0fa7fb15a1d2c1e44db2)
+  has_many :comments, dependent: :destroy
```
Comment (DaichiSaito) [inline app/models/post.rb:33]: > deleteの場合、子であるcommentsまで削除されないので要注意

子も削除されますよ。ただActiveRecordを介さないのでバリデーション等々が走らなくなります。

File: app/models/user.rb [hunk 5]
```diff
@@ -28,6 +28,9 @@ class User < ApplicationRecord
   # [Active Record の関連付け \- Railsガイド](https://railsguides.jp/association_basics.html#belongs-to%E3%81%AE%E3%82%AA%E3%83%97%E3%82%B7%E3%83%A7%E3%83%B3-dependent)
   has_many :posts, dependent: :destroy
 
+  # Userは消えてもコメントだけは残しておきたい場合、nullifyオプションを使うとよい(その場合、nullifyはhas_oneの方に使う)
```
Comment (DaichiSaito) [inline app/models/user.rb:31]: > その場合、nullifyはhas_oneの方に使う

どういうことだろ？？

File: app/views/comments/create.js.slim [hunk 6]
```diff
@@ -0,0 +1,16 @@
+/ NGワードを含む場合
+- if @ng_flag
+  / NGワードを含む投稿をしたのでアラートを出す
+  | alert("NGワードが含まれています。綺麗な言葉を使いましょう。");
+  / NGワードを含む投稿をしたので、次の投稿がしやすいように投稿欄を空にする
+  | $('.input-comment-body').val('');
+/ バリデーションに抵触した場合
+- elsif @comment.errors.present?
```
Comment (DaichiSaito) [inline app/views/comments/create.js.slim:8]: モデルのバリデーションエラーに含められたら綺麗に実装できますよね〜。まぁ時間かけすぎるのももったいないので今のままいってみましょーか！

File: config/locales/my_swears.yml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+regex:
+  '^(?=.*おっぱい).*$': ["sexual"]
+  '^(?=.*キチガイ).*$': ["inappropriate"]
+  '^(?=.*デブ).*$': ["discriminatory"]
+  '^(?=.*ハゲ).*$': ["discriminatory"]
+  '^(?=.*バカ).*$': ["insult"]
+  '^(?=.*死ね).*$': ["insult"]
```
Comment (DaichiSaito) [inline config/locales/my_swears.yml:7]: 笑

File: app/models/post.rb (hunk 4 above)
Comment (miketa-webprgr) [reply app/models/post.rb:33]: > ActiveRecordを介さない

違いはここだけなんですね！
訂正ありがとうございます！

deleteの場合、バリデーションが走らなくて困る場合というのはあるんでしょうか？

File: app/models/user.rb (hunk 5 above)
Comment (miketa-webprgr) [reply app/models/user.rb:31]: 不勉強であるが故に書いてしまった適当コメントですね。。。
その時は何を思ったのか、has_manyには使えないと勘違いしちゃいました

File: app/views/comments/create.js.slim (hunk 6 above)
Comment (miketa-webprgr) [reply app/views/comments/create.js.slim:8]: そうですよね、バリデーションエラーに含めることができれば、このコードも不要になりますし。
とりあえず、今のままで行かせていただきます！

ただ、この話は別にして、モデルのカスタムバリデーションはマスターしたいと思いました！
少し見た限り、そんなに難しそうではないんですけどね。。。

File: config/locales/my_swears.yml (hunk 7 above)
Comment (miketa-webprgr) [reply config/locales/my_swears.yml:7]: コメントすると目立つんでやめてください笑
恥ずかしくて、このあたりが限界でした笑

File: app/models/post.rb (hunk 4 above)
Comment (DaichiSaito) [reply app/models/post.rb:33]: user has_many questions

question has_many answers

という関係があったとして（teratailみたいなイメージ）、例えば回答がついてる質問は消せないというバリデーションがquestionについてるとします。
その時 user has_many questions, dependent: :delete_all だと回答がついてる質問も全部消えちゃいます！

File: app/controllers/comments_controller.rb (hunk 3 above)
Comment (DaichiSaito) [reply app/controllers/comments_controller.rb:48]: 簡単アプリの相互レビュー会！

File: app/models/post.rb (hunk 4 above)
Comment (miketa-webprgr) [reply app/models/post.rb:33]: なるほど！
分かりやすい説明ありがとうございます！！
